### ISSUE TITLE
Delays when some nodes are down

### ISSUE THREAD

**Issue by stek29** [minio/minio#9399]
[Originally reported on Slack](https://minio.slack.com/archives/C3NDUB8UA/p1587391420182900)

1. Set up distributed minio 4 node with 4 disks per node (see docker-compose file in the end)
2. Bring one node down _while keeping its dns record alive_ (i.e. on network level):
```sh
# docker-compose exec minioX sh
apk add iptables
iptables -A INPUT --proto tcp --dport 9000 -j DROP
iptables -A OUTPUT --proto tcp --dport 9000 -j DROP
```
3. Observe (non-configurable!) timeouts using mc:
- mc mb, rb, ls work fine
- mc cp local-file minio hangs for 1 minute
- mc cp minio local-file hangs for 1 minute
- mc admin info hangs for 3 minutes

## Expected Behavior
No timeout for read operations, or at least timeout less than 1 minute.

## Current Behavior
```
/ # time mc cp test.txt h1/time/test.txt
real    1m 0.21s
user    0m 0.02s
sys     0m 0.03s
/ # time mc cp h1/time/test.txt test2.txt
real    1m 0.17s
user    0m 0.01s
sys     0m 0.04s
/ # time mc admin info h1
●  minio2:9000
   Uptime: offline

●  minio3:9000
   Uptime: 7 hours 
   Version: 2020-04-15T19:42:18Z
   Network: 3/4 OK 
   Drives: 4/4 OK 

●  minio4:9000
   Uptime: 7 hours 
   Version: 2020-04-15T19:42:18Z
   Network: 3/4 OK 
   Drives: 4/4 OK 

●  minio1:9000
   Uptime: 7 hours 
   Version: 2020-04-15T19:42:18Z
   Network: 3/4 OK 
   Drives: 4/4 OK 

210 MiB Used, 2 Buckets, 7 Objects
12 drives online, 0 drives offline
real    3m 0.16s
user    0m 0.01s
sys     0m 0.02s
```

## Possible Solution
Refactor timeouts spread through the codebase to make them configurable via environment variables.
Decrease download timeout.

## Steps to Reproduce (for bugs)
See above

## Context
I want to set up baremetal highly-available minio cluster, which should continue working if some nodes are down.
minio already allows me to do that, but high timeouts make cluster completely unusable. I can't call it working if it takes 1 minute to download a 4 byte file.

## Regression
Maybe -- see #7400

## Your Environment

* Version used (`minio version`): 2020-04-15T19:42:18Z
<details><summary>Docker Compose</summary>
version: "3.7"

x-minio: &minio
  image: minio/minio:RELEASE.2020-04-15T19-42-18Z
  environment:
    MINIO_ACCESS_KEY: minio
    MINIO_SECRET_KEY: minio123ty
  cap_add:
    - NET_ADMIN
  command: server http://minio{1...4}/data{1...4}
  healthcheck:
    test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
    interval: 30s
    timeout: 20s
    retries: 3

# starts 4 docker containers running minio server instances. Each
# minio server's web interface will be accessible on the host at port
# 9001 through 9004.
services:
  minio1:
    <<: *minio
    volumes:
      - data1-1:/data1
      - data1-2:/data2
      - data1-3:/data3
      - data1-4:/data4
    ports:
      - "9001:9000"

  minio2:
    <<: *minio
    volumes:
      - data2-1:/data1
      - data2-2:/data2
      - data2-3:/data3
      - data2-4:/data4
    ports:
      - "9002:9000"

  minio3:
    <<: *minio
    volumes:
      - data3-1:/data1
      - data3-2:/data2
      - data3-3:/data3
      - data3-4:/data4
    ports:
      - "9003:9000"

  minio4:
    <<: *minio
    volumes:
      - data4-1:/data1
      - data4-2:/data2
      - data4-3:/data3
      - data4-4:/data4
    ports:
      - "9004:9000"

volumes:
  data1-1:
  data1-2:
  data1-3:
  data1-4:
  data2-1:
  data2-2:
  data2-3:
  data2-4:
  data3-1:
  data3-2:
  data3-3:
  data3-4:
  data4-1:
  data4-2:
  data4-3:
  data4-4:
</details>

Similar to #7618 and #7400.

**Comment by stek29** [comment]
Actually, it looks like #7400 duplicate -- I found that issue only when I finished making this one.
However, I'm not quite sure how it was fixed by op, and I'd still love to have timeouts configurable.

update: according to #7400, only first request after node being down should take 1 minute. however, I am not able to reproduce that: retrieving same file twice from minio takes 1 minute each time.

**Comment by stek29** [comment]
I've tested with `RELEASE.2019-03-20T22-38-47Z`, released just after #7398, and:
- first download hangs for 5 minutes
- further downloads dont hang
- first upload hangs for 2 minutes
- further uploads dont hang
- first admin info is broken with io timeouts each time.

so, **it is a regression**. I'm not sure when it was introduced, it was somewhere between `RELEASE.2019-03-20T22-38-47Z` and `2020-04-15T19:42:18Z`.

**Comment by harshavardhana** [comment]
thanks we shall take a look @stek29

**Comment by harshavardhana** [comment]
closing as a duplicate of #9508

**Comment by stek29** [comment]
@harshavardhana im not sure it’s a duplicate.
1. This issue is older
2. It has more info
3. It’s assigned to a milestone unlike #9508 
3. It’s different — time values are different, it’s not about which node the client is connected to — it’s about success after 1 minute, not about failure after 25 minutes.

Also, maybe I should try to dissect which release introduced the issue?

**Comment by harshavardhana** [comment]
> @harshavardhana im not sure it’s a duplicate.
> 1. This issue is older
> 2. It has more info
> 3. It’s assigned to a milestone unlike #9508 
> 3. It’s different — time values are different, it’s not about which node the client is connected to — it’s about success after 1 minute, not about failure after 25 minutes.
> 
> Also, maybe I should try to dissect which release introduced the issue?

It is duplicate because the underlying problem is same. I closed this issue because I know where the problem is and why they are related.
